mapeditor/mapview.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <QGraphicsScene>
+#include <QGraphicsView>
+#include "scenelayer.h"
+#include "../lib/int3.h"
+
+
+class CGObjectInstance;
+class MainWindow;
+class MapController;
+
+class MapSceneBase : public QGraphicsScene
+{
+	Q_OBJECT;
+public:
+	MapSceneBase(int lvl);
+	
+	const int level;
+	
+	virtual void updateViews();
+	virtual void initialize(MapController &);
+	
+protected:
+	virtual std::list<AbstractLayer *> getAbstractLayers() = 0;
```

Comment:
Why use `std::list`? It looks like you're just iterating through the list, and not removing items from the middle of it. `std::vector` would offer better performance here; better cache locality.

Reply:
Actually, not quite, even about performance. Vector has O(NlnN) complexity for N sequential push backs, while list has O(N). But the reason I chose list was not about performance of course (such kind of performance improvements are completely useless here).
I use list if I want implicitly forbid random access for container. As soon as you have random access you have potential risk of access violation (out of range in the best case) so everything which might happen - happens soon or late:)

Reply:
That makes sense. In that case, we need to use a std::list.

Reply:
Afaik vector should have amortized O(N) and I have seen an article showing that linked list is way slower than vector of ints because it needs allocation for each new item while vector always allocates 2x capacity each time it has not enough space and just does memory copy. So if the vector item is quite big - this might be a problem. In this case we can store shared pts as an option.
But I agree regarding random access)